### PR TITLE
Squash allocations in selector parsing and other hotspots

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Noticed that selector parsing was a memory allocation hotspot in clusters with lots of policy.  Most of that was down to _validating_ selectors so we were just throwing away the parsed selector.

- Check log level in various places before logging to avoid allocs in logrus.
- Add `tokenizer.AppendTokens()` to do tokenization into a pre-allocated buffer.  Use a shared instance of the parser to hold the shared token buffer (protected my mutex).
- Add dedicatied Validate() function that calls parser with a flag telling it not to allocate the selector.
- Replace regexes with custom code.  The regex engine is relatively slow for these simple matches and it allocates for every call.
- Modernise some of the tokenizer code using `strings.Cut` and friends; makes for easier reading and less string slicing.
- Generate a String() method for the token constants.
- Add more coverage tests.

Before:
```
goos: linux
goarch: amd64
pkg: github.com/projectcalico/calico/libcalico-go/lib/selector
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkParse-12    	  162223	      6416 ns/op	    2633 B/op	      46 allocs/op
PASS
```

After:
```
goos: linux
goarch: amd64
pkg: github.com/projectcalico/calico/libcalico-go/lib/selector
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkParse-12       	  919357	      1209 ns/op	     520 B/op	      22 allocs/op
BenchmarkValidate-12    	 3878425	       293.8 ns/op	       0 B/op	       0 allocs/op
PASS
```
## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Improve performance of selector parsing and validation.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by mazdakn — APPROVED** [review]
LGTM. Just left a few questions and nits.

File: libcalico-go/lib/selector/tokenizer/tokenizer.go
Comment (fasaxc) [reply libcalico-go/lib/selector/tokenizer/tokenizer.go:80]: I think TrimLeft was allocating/doing work to calculate the cut set.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, squash the commits when I merge it, and leave the branch after I've merged it.

**Pull request merged**
